### PR TITLE
frontend: move project selector to strict typescript

### PR DESCRIPTION
**Pull request by dschaller**
### Description

Enable strict typescript rules for project selector and make appropriate fixes.

### Testing Performed

manual

### CODE REVIEW

File: frontend/workflows/projectSelector/src/dash-hooks.tsx [hunk 1]
```diff
@@ -5,7 +5,7 @@ import type { DashAction, DashState } from "./types";
 export const DashStateContext = React.createContext<DashState | undefined>(undefined);
 
 export const DashDispatchContext = React.createContext<(action: DashAction) => void | undefined>(
-  undefined
+  () => undefined
 );
```
Comment (danielhochman) [inline frontend/workflows/projectSelector/src/dash-hooks.tsx:9]:
```suggestion
export const DashDispatchContext = React.createContext<((action: DashAction) => void) | undefined>(
  undefined
);
```

File: frontend/workflows/projectSelector/src/helpers.tsx [hunk 2]
```diff
@@ -16,17 +16,15 @@ const useReducerState = () => {
   return React.useContext(StateContext);
 };
 
-const DispatchContext = React.createContext<(action: Action) => void | undefined>(undefined);
+const DispatchContext = React.createContext<(action: Action) => void | undefined>(() => undefined);
```
Comment (danielhochman) [inline frontend/workflows/projectSelector/src/helpers.tsx:19]:
```suggestion
const DispatchContext = React.createContext<((action: Action) => void) | undefined>(undefined);
```

File: frontend/workflows/projectSelector/src/project-group.tsx [hunk 3]
```diff
@@ -116,8 +116,9 @@ const ProjectGroup: React.FC<ProjectGroupProps> = ({ title, group, displayToggle
 
   const [collapsed, setCollapsed] = React.useState(false);
 
-  const numProjects = Object.keys(state[group]).length;
-  const checkedProjects = Object.keys(state[group]).filter(k => state[group][k].checked);
+  const groupKeys = Object.keys(state?.[group] || {});
```
Comment (danielhochman) [inline frontend/workflows/projectSelector/src/project-group.tsx:119]: nit

```suggestion
  const groupKeys = Object.keys(state?.[group] ?? {});
```

File: frontend/workflows/projectSelector/src/project-group.tsx [hunk 4]
```diff
@@ -153,53 +154,51 @@ const ProjectGroup: React.FC<ProjectGroupProps> = ({ title, group, displayToggle
...
-                  name={key}
-                  size="small"
-                  disabled={state.loading}
-                  onChange={() =>
+          {groupKeys.sort().map(key => (
+            <StyledMenuItem key={key}>
+              <Checkbox
+                name={key}
+                size="small"
+                disabled={state?.loading}
```
Comment (danielhochman) [inline frontend/workflows/projectSelector/src/project-group.tsx:162]: i think you can eliminate the added `state?` checks by throwing an error from the hook and never returning an undefined state, which should never happen

**Review by danielhochman — COMMENTED** [review]
first round

File: frontend/workflows/projectSelector/src/project-group.tsx (hunk 4 above)
Comment (dschaller) [reply frontend/workflows/projectSelector/src/project-group.tsx:162]: If we don't expect state to ever be undefined should we change the initialization of it (and types) to not allow for undefined?

Comment (scarlettperry) [reply frontend/workflows/projectSelector/src/project-group.tsx:163]: @dschaller , just for my understanding, do you mean something like this?
```
const StateContext = React.createContext({
  [Group.PROJECTS]: {},
  [Group.UPSTREAM]: {},
  [Group.DOWNSTREAM]: {},
  projectData: {},
  loading: false,
  error: undefined,
} as State);
```
we also initialize it with default values in the project selector component in `useReducer`

**Review by scarlettperry — APPROVED** [review]

File: frontend/workflows/projectSelector/src/project-group.tsx (hunk 4 above)
Comment (dschaller) [reply frontend/workflows/projectSelector/src/project-group.tsx:163]: Chatted IRL with @scarlettperry and we are going to punt on removing undefined as a type until later.

**Review by mikecutalo — APPROVED** [review]

**Pull request merged**